docs/src/proposals/multi-writers.md

```diff
@@ -0,0 +1,188 @@
+# Enable Multi-Writers in Shred Insertion
+
+## Problem
+The current shred insertion is single-threaded which prevents us from fully
+utilizing the write bandwidth of our RocksDB backed BlockStore.
+
+The entire shred insertion is protected by an `insert_shreds_lock`, which
+ensures the atomicity of the read-modify-write operations on meta-data
+column families.
+
+## Opportunities
+RocksDB has an API called merge operator.  In a high level, it offers
+lockless read-modify-write operation where we only write the delta
+as a merge operand to the DB, and such merge operand will be used to
+compute the final value by merging it with its older value.
+
+For instance, suppose we want to read a value and write value+1 back
+to RocksDB.  Instead of obtaining a write lock, reading the value,
+performing value + 1, writing value + 1 back to the DB then release
+the write lock, we simply write `+1` back to the DB and let such `+1`
+being processed only at the read time or compaction time.  This allows
+us to update an entry without the need of acquiring a write lock.
+
+## Existing Solution
+A write lock is used to perform read-modify-write during the shred insertion,
+which prevents us from having multiple writers.
+
+## Proposed Solution
+To enable running multiple writers in parallel, we need to ensure the
+atomicity of each shred insertion and avoid data races.  To achieve this,
+both the shared in-memory structures and BlockStore access have to be
+thread-safe.
+
+For the shared in-memory structures such as BlockstoreInsertionMetrics,
+we need to use a thread-safe version or only update them all at once
+at the end of the function call to minimize the lock duration.
+
+For the BlockStore, we will utilize RocksDB's merge operator to perform
+read-modify-write operation to avoid data race between different writers.
+
+Specifically, the `insert_shred()` function touches the following column
+families:
+
+### DeadSlots
+DeadSlots stores [(slot, flag)](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L1212)
+pairs indicating whether a particular slot can be safely ignored / deleted.
+As it does not involve in any read-modify-write based on its previous state,
+simply using write-batch with put operation is enough to ensure its atomicity
+and thread-safety.
+
+### ShredData / ShredCode
+Both [ShredData](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L1473-L1478)
+and [ShredCode](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L1274)
+store data directly from the received shred and do not rely on any previous
+metadata.  As a result, they are simply write opreations which do not require
+merge operation to ensure potential read-modify-write atomicity.
```

Comment:
Only concerns I would have here:

1) Races with `clear_unconfirmed_slot()`: https://github.com/solana-labs/solana/blob/45e56c599d726766f7c4440a73ff29b77fa5abf6/ledger/src/blockstore.rs#L980-L982, which is removing the slot data and coding shreds. If a deletion is issued, this will not cause any corruption right? If a deletion happens during the middle of a merge, worst case the merge is just writing stale data (this is probably acceptable)?

2) Will the merge updates to the column families happen in an atomic transaction? Temporary inconsistency between column families might break some inconsistencies. For instance if the SlotMeta merge happens first, we may have a "consumed" value of `N`, but the shred for `N` may not have been written to blockstore yet. Or perhaps the index hasn't been updated yet for that slot, which would cause code like this to panic: https://github.com/solana-labs/solana/blob/45e56c599d726766f7c4440a73ff29b77fa5abf6/core/src/repair_generic_traversal.rs#L59-L61

3. Also, the validator may crash during the merge operation for one of these column families. For instance if we do the merge operation on the SlotMeta first, and then we crash during the merge of the ShredData, these two column families will be out of sync on restart. Will we need some consistency checkers?